### ISSUE TITLE
date/time stamping for future releases.

### ISSUE THREAD

**Issue by proseLA** [zencart/zencart#2028]
is there any way to move the time date stamping within the code so that year comes first and perhaps remove the day of the week?  ie, a current time stamp on the file in v156:

``` * @version $Id: dakanji Thu Jun 14 20:55:08 2018 +0300 Modified in v1.5.6 $ ```

would be better suited if something more like:

``` * @version $Id: dakanji 2018 Jun 14 20:55:08 +0300 Modified in v1.5.6 $```

i find the year the most important, and the day of the week useless.... 

it does seem like there has been an evolution of stamping as here is another one i see that does not include the year:

```* @version $Id: Author: zcwilt Fri Apr 15 Modified in v1.5.5 $```

would be very helpful when doing merges....  at least for me.

**Comment by Zen4All** [comment]
And match the line endings to the document. Now the line endings are linux by default, which makes a lot of files with dos line endings end up as a mixed line endings, making comparing files a nightmare.

**Comment by proseLA** [comment]
version stamp on admin/includes/auto_loaders/config.core.php

`* @version $Id: mc12345678 Sun Jan 20 01:27:39 2019 -0400 Modified in v1.5.6b $`

**Comment by zcwilt** [comment]
tl;dr We have written custom code to do this.

The longer answer is that we use git diff to get a list of all files changed between 2 git hashes
from that we know new/modified files
we then loop through those files and use git log to get the timestamps of the last change/author etc.
we can then use all that information to re-write the header information.

**Comment by drbyte** [comment]
... but we only run the script against the files when we're readying for release.
So any commits made since the last release may have un-updated headers, or incorrectly-formatted headers, or incorrect dates/versions in them.

So in the case of @proseLA's last comment about someone's commit headers being wrong, ignore it until an official release is made. We'll stamp them uniformly at that time.

**Comment by barco57** [comment]
apparently, this procedure does not always work...
as an example, extras/curltester.php has had the same timestamps in the file since 1.5.5 but there were changes to the file between 1.5.5a-b and again 1.5.5e-f

I noticed this the other day when troubleshooting a curl issue and testing on different servers, and I grabbed copied of curltester.php from different versions for some reason, and a quick glance at the time stamps tells me they are the same file, but upon closer inspection, I realized that they were different.

**Comment by drbyte** [comment]
Hmmmmm .... Maybe the stamper script skipped the `extras` directory.

**Comment by zcwilt** [comment]
I don't think we have ever omitted the extras directory from version stamping.
And in fact https://github.com/zencart/zencart/commit/8f7c0e0e9b9c5ba1f3b3a301fb81e6d0e62f9148#diff-73e8942af77caf8001be3cbbe8ff1e22 shows us version stamping ``curltester``

What the history does show is that during the 1.5.5  release cycles is that we only ever used v1.5.5 as a version stamp and not say v1.5.5a, v1.5.5b etc so there are probably lots more files that were changed during v1.5.5-.v1.5.5f that don't have a correct version stamp

I'm really not sure why we did that.

**Comment by zcwilt** [comment]
A bit of history first.

Before moving to GitHub we used svn to manage our code repositories.
svn allowed for automatic expansion of lines in the header, based on replacement of certain tags
This meant that commits would automatically update the header lines putting in correct dates and commit author.
At that time we didn't add the modified  in version, new in version info.
plus as well  our source was closed so only team members had commit access. no pull requests.

When we moved to using GitHub we had a choice to make. Either stop using the $ID$ line or find a new way to provide that information.
There was some discussion to not even present this information, but it was decided that it was useful for those wanting to see diffs between versions directly rather than just relying on the documentation showing new/modified files.


In a perfect world we could have relied on committers to update that header information, but the review process for that would be onerous, especially where commits might touch multiple files.

Instead we decided to write some custom code to scan all commits between branches to find files that were new or modified and update headers based on the information that git diff/git log could provide.

The original code had some downsides. The biggest was that the code could only be run once. So once we had run version stamping, we couldn't run it again. If we had to make changes after version stamping we had to make sure we manually changed the header information.

The reason for this is that the version stamping process itself adds commits and those commits will be tagged with the committer who merges the version stamp branch. Running the version stamping process again would change the author/dates to the author of that merge/commit.

So given all that history, lets try and answer your questions.

> are you saying that with some merged PRs, contributors are manually modifying stamps and until an official release, those modified stamps "may" indicate a changed file since the last official release?

Yes. some committers may modify the header info, some may not. That means you cannot rely on the truth of the headers between releases.

> more importantly, i'm curious about this commit 23875d3. v156 was released 12/10/2018. v156a was released 1/4/2019. any file that was in the above referenced commit was changed in between v156 and v156a? (not just the copyright date in the header.)

Yes, the version stamping process only looks at new/changed files, so any file that was not changed would never have it's header info changed. This may mean that some files will have old copyright dates. There are 2 reasons for this. Firstly copyright dates are not legally binding. We could just not put a copyright date in there. Also changing copyright dates for all files would produce horrible diffs.


> looking at the version stamping of admin/languages, the last version stamp commit would indicate an older working copy with the same code that was still somehow current??

Not sure what you are referencing here. Do you mean admin/languages.php Although it does look as though the author naming has gone awry. Going to have to look why

> does the official release stamp always stay with 1 particular user for that release? or has the policy for putting an ID on an official release changed over the years as well.

Do you mean what author it uses for the release stamp ? 
It's always supposed to use the person who made the last commit to the file. Now however I'm of the opinion that our version stamping does not always pick the correct author for some reason. Although it's only in a few edge cases.

How are you using version stamping in your process?. e.g do you rely on it for diffing between versions.
What are other people using the version stamping for?

**Comment by proseLA** [comment]
- thanks for all of the background info.  very interesting and enlightening.
- i was mistaken in my comment on `admin/languages.php`.  i thought there was an inconsistency in the stamp from one of the contributors which confused me.  that was my mistake.
- unfortunately during an upgrade, i need to modify far too many core files based on the customizations i have done for some clients.  so i am continuously using meld/diff to examine diffs.  i carry over timestamps onto my modified code, so that i can tell when i last looked at that core file compared to where that file is now.

again, thanks for the info.  it helps in my understanding of your process.

best.
